File: next element with greater freequency.py
class Solution:
    def print_next_greater_freq(self,arr,n):
        dict={}
        for i in range(n):
            if arr[i] not in dict:
                dict[arr[i]]=1
            else:
                dict[arr[i]]+=1
        frq=[0]*n
        for i in range(n):
            frq[i]=dict[arr[i]]
        st=[n-1]
        ans=[-1]*n
        #print(frq)
        for i in range(n-2,-1,-1):
            #print(st)
            if frq[i]<frq[st[-1]]:
                ans[i]=arr[st[-1]]
                st+=[i]
            else:
                while len(st) and frq[i]>=frq[st[-1]]:
                    st.pop()
                if len(st)==0:
                    ans[i]=-1
                else:
                    ans[i]=arr[st[-1]]
                st+=[i]
        return ans

File: next element with greater freequency.cpp
public:
    vector<int> print_next_greater_freq(int arr[],int n)
    {
        map<int,int>dict;
        for(int i=0;i<n;i++)
        {
          dict[arr[i]]+=1;
        }
        vector<int>frq(n,0);
        for(int i=0;i<n;i++)
        {
            frq[i]=dict[arr[i]];
            //cout<<dict[arr[i]]<<' ';
        }
        stack<int>st;
        st.push(n-1);
        vector<int>ans(n,-1);
        for(int i=n-2;i>-1;i--)
        {
            if(frq[i]<frq[st.top()])
            {
                ans[i]=arr[st.top()];
                st.push(i);
            }
            else
            {
                while(!st.empty() and frq[i]>=frq[st.top()])
                    st.pop();
                if(st.empty())
                    ans[i]=-1;
                else
                    ans[i]=arr[st.top()];
                st.push(i);
            }
        }
        return ans;
        
        
    }
};
